SmartDeviceLink/SDLSecondaryTransportManager.m

```diff
@@ -0,0 +1,806 @@
+        return SDLTransportSelectionIAP;
+    } else {
+        // "SPP_BLUETOOTH" and "AOA_USB" are not used
+        return SDLTransportSelectionDisabled;
+    }
+}
+
+- (nullable NSArray<SDLTransportClassBox *> *)sdl_convertServiceTransports:(nullable NSArray<NSNumber *> *)transports {
+    if (transports == nil) {
+        return nil;
+    }
+
+    // Actually there is nothing to "convert" here. We just check the range and recreate an array.
+    NSMutableArray<SDLTransportClassBox *> *array = [NSMutableArray new];
+    for (NSNumber *num in transports) {
+        int transport = [num intValue];
+        if (transport == 1) {
+            [array addObject:@(SDLTransportClassPrimary)];
+        } else if (transport == 2) {
+            [array addObject:@(SDLTransportClassSecondary)];
+        } else {
+            SDLLogV(@"Unknown transport class received: %d", transport);
+        }
+    }
+
+    return [array copy];
+}
+
+- (SDLTransportSelection)sdl_getTransportSelectionFromProtocol:(SDLProtocol *)protocol {
+    if ([protocol.transport isMemberOfClass:[SDLIAPTransport class]]) {
+        return SDLTransportSelectionIAP;
+    } else if ([protocol.transport isMemberOfClass:[SDLTCPTransport class]]) {
+        return SDLTransportSelectionTCP;
+    } else {
+        SDLLogD(@"Unknown type of primary transport");
+        return SDLTransportSelectionDisabled;
+    }
+}
+
+
+#pragma mark - TransportEventUpdate notification implementation
+
+// called from SDLProtocol's _receiveQueue of "primary" transport
+- (void)sdl_onTransportEventUpdate:(NSNotification *)notification {
+    NSAssert([notification.userInfo[SDLNotificationUserInfoObject] isKindOfClass:[SDLControlFramePayloadTransportEventUpdate class]], @"Unexpected type of object in the notification");
+
+    SDLControlFramePayloadTransportEventUpdate *payload = notification.userInfo[SDLNotificationUserInfoObject];
+
+    dispatch_async(_stateMachineQueue, ^{
+        BOOL updated = NO;
+
+        if (payload.tcpIpAddress != nil) {
+            if (![self.ipAddress isEqualToString:payload.tcpIpAddress]) {
+                self.ipAddress = payload.tcpIpAddress;
+                updated = YES;
+                SDLLogD(@"TCP transport IP address updated: %@", self.ipAddress);
+            }
+        }
+        if (payload.tcpPort != SDLControlFrameInt32NotFound) {
+            if (self.tcpPort != payload.tcpPort) {
+                self.tcpPort = payload.tcpPort;
+                updated = YES;
+                SDLLogD(@"TCP transport port number updated: %d", self.tcpPort);
+            }
+        }
+
+        if (updated) {
+            [self sdl_handleTransportEventUpdate];
+        }
+    });
+}
+
+
+#pragma mark - Transport management
+
+// try establishing secondary transport. Returns NO if failed
+- (BOOL)sdl_connectSecondaryTransport {
+    if (self.secondaryTransport != nil) {
+        SDLLogW(@"Secondary transport is already started.");
```

Comment:
This isn't enough detail. Maybe "Attempting to connect secondary transport, but it's already started"

Reply:
Thanks, updated in commit 5009bc7f0137147a9211bbf3725e2dee7b5cba98.